x-pack/plugin/esql/arrow/src/main/java/org/elasticsearch/xpack/esql/arrow/BlockConverter.java

```diff
@@ -0,0 +1,477 @@
+    private final String esqlType;
+
+    protected BlockConverter(String esqlType, Types.MinorType minorType) {
+        // Add the exact ESQL type as field metadata
+        var meta = Map.of("elastic:type", esqlType);
+        this.fieldType = new FieldType(true, minorType.getType(), null, meta);
+        this.esqlType = esqlType;
+    }
+
+    public final String esqlType() {
+        return this.esqlType;
+    }
+
+    public final FieldType arrowFieldType() {
+        return this.fieldType;
+    }
+
+    protected final <T extends Block> T castBlock(Block block, Class<T> clazz) {
+        if (clazz.isAssignableFrom(block.getClass())) {
+            return clazz.cast(block);
+        } else {
+            throw new IllegalArgumentException(
+                "Field type ["
+                    + esqlType
+                    + "] expects a ["
+                    + clazz.getSimpleName()
+                    + "] but we got a ["
+                    + block.getClass().getSimpleName()
+                    + "]"
+            );
+        }
+    }
+
+    // Block.nullValuesCount was more efficient but was removed in https://github.com/elastic/elasticsearch/pull/108916
+    protected int nullValuesCount(Block block) {
+        if (block.mayHaveNulls() == false) {
+            return 0;
+        }
+
+        if (block.areAllValuesNull()) {
+            return block.getPositionCount();
+        }
+
+        int count = 0;
+        for (int i = 0; i < block.getPositionCount(); i++) {
+            if (block.isNull(i)) {
+                count++;
+            }
+        }
+        return count;
+    }
+
+    public interface BufWriter {
+        long write(RecyclerBytesStreamOutput out) throws IOException;
+    }
+
+    /**
+     * Convert a block into Arrow buffers.
+     * @param block the ESQL block
+     * @param bufs arrow buffers, used to track sizes
+     * @param bufWriters buffer writers, that will do the actual work of writing the data
+     */
+    public abstract void convert(Block block, List<ArrowBuf> bufs, List<BufWriter> bufWriters);
+
+    /**
+     * Conversion of Double blocks
+     */
+    public static class AsFloat64 extends BlockConverter {
+
+        public AsFloat64(String esqlType) {
+            super(esqlType, Types.MinorType.FLOAT8);
+        }
+
+        @Override
+        public void convert(Block b, List<ArrowBuf> bufs, List<BufWriter> bufWriters) {
+            DoubleBlock block = castBlock(b, DoubleBlock.class);
+
+            accumulateVectorValidity(bufs, bufWriters, block);
+
+            bufs.add(dummyArrowBuf(vectorLength(block)));
+            bufWriters.add(out -> {
+                if (block.areAllValuesNull()) {
+                    return BlockConverter.writeZeroes(out, vectorLength(block));
+                }
+
+                // TODO could we "just" get the memory of the array and dump it?
+                int count = block.getPositionCount();
+                for (int i = 0; i < count; i++) {
```

Comment:
I added explicit rejection of multi-valued arrays (for now) in c7cd12d026558c73b1c847856fcf82551bbc9a16 so this is ok as we only have single-valued arrays.

Null is handled, based on the observation that a null value uses a position and its value is zero in number blocks. I don't know if this part of the contract (I added a test for it), but it should, so that we can optimize processing of arrays where `block.mayHaveMultivaluedFields() == true`.

The validity bitmap is also output in `BlockConverter.accumulateValidity()`.

That will of course have to be revisited when handling mutl-values.